### PR TITLE
[SPARK-15203][Deploy]The spark daemon shell script error, daemon process start successfully but script output fail message.

### PR DESCRIPTION
**Pull request by WeichenXu123**
This bug is because, sbin/spark-daemon.sh script use bin/spark-class shell to start daemon, then sleep  2s and check whether the daemon process exists, using shell script like following:
if [[ ! $(ps -p "$newpid" -o comm=) =~ "java" ]]
the problem is, some machine with bad performance may start the daemon using a long time(exceeding 2s), but still can start daemon successfully, but in this case, the shell script judgement  ! $(ps -p "$newpid" -o comm=) =~ "java"  will fail, because at this time, the $newpid process is still shell process, until the daemon started, it turns into java process. That's the reason cause the bug.
In order to reproduce the bug more easily, you can change 
sleep 2  ==>  sleep 0.01 (sbin/spark-daemon.sh, line 165)

to fix this bug, 
I replace
 ! $(ps -p "$newpid" -o comm=) =~ "java"   (sbin/spark-daemon.sh, line 167)
with 
-z $(ps --no-headers -p "$newpid")

### CODE REVIEW

File: sbin/spark-daemon.sh [hunk 1]
```diff
@@ -164,7 +164,7 @@ run_command() {
   echo "$newpid" > "$pid"
   sleep 2
   # Check if the process has died; in that case we'll tail the log so the user can see
-  if [[ ! $(ps -p "$newpid" -o comm=) =~ "java" ]]; then
+  if [[ -z $(ps --no-headers -p "$newpid") ]]; then
```
Comment (srowen) [inline sbin/spark-daemon.sh]: I don't see the option `--no-headers` on OS X, so I suspect it could be a BSD vs Gnu thing. It will fail though.

I don't see how this change addresses the problem though. It still occurs 2 seconds after the start, but it's not checking whether the java process started. Do you mean some kind of loop here, like checking 3 times?

What's the scenario where the bash script takes a long time to run?

**Comment by WeichenXu123** [comment]
@srowen 
because spark-daemon.sh using exec command to start the java daemon process.
when run script, the spark-daemon.sh process will exists for a little time(stage1) and then changed to a java process(stage2), then if java process throw exception, it will die.
if stage1 spending more then 2s, then the judgement if [[ ! $(ps -p "$newpid" -o comm=) =~ "java" ]]; is fail, it will regard the process as failed but in fact it has not come into stag2. In my virtual machine, this thing offen happen.

**Comment by srowen** [comment]
Hm, how would stage 1 take any significant time though? I'm missing what could take that much time.
The problem with this change is that it fails to check whether the java process started, so we can't do this. At best a loop to check the status every second for several seconds or something.

**Comment by WeichenXu123** [comment]
@srowen At my virtual machine, After OS started and start spark daemon, the stage 1 describe above will took a long time,often exceeding 2s, I think the java daemon start need to load a huge number of jars so it will took several time when started at the first time.

**Comment by srowen** [comment]
At that point the Java process is already running though. I want to make sure we are solving the problem you think.

**Comment by WeichenXu123** [comment]
@srowen Er...I am also a little strange while stage 1 may took a long time but it really happen several times... If there is time I will do a more detailed test.

**Comment by srowen** [comment]
I think the way forward is to make a different change, if you're still pretty sure this is an issue: try this a small number of times in a loop until it succeeds or the max iterations is reached. Same check, just tried once a second for 5 seconds or something.

**Comment by WeichenXu123** [comment]
@srowen According to your suggestion, I add a loop to check whether it pass STAGE-1 and launch java daemon. And I recover the check statement `! $(ps -p "$newpid" -o comm=) =~ "java"`

File: sbin/spark-daemon.sh [hunk 2]
```diff
@@ -162,6 +162,15 @@ run_command() {
   esac
 
   echo "$newpid" > "$pid"
+
+  for i in {1..10}
+  do
+    if [[ $(ps -p "$newpid" -o comm=) =~ "java" ]]; then
+       break
+    fi
+    sleep 0.5
```
Comment (srowen) [inline sbin/spark-daemon.sh]: Maybe once a second for 5 seconds? you need to remove the "sleep 2" below.
Maybe you can use a variable rather than re-test the condition after the loop?

However, the PR has also been messed up with some unrelated changes. Can you fix or else open a new PR?

**Comment by WeichenXu123** [comment]
In order to check this potential problem more carefully, We can add the following test code like this:

`
echo "$newpid" > "$pid"
  **for i in {1..30};do
    echo "check daemon stage $i" `ps -p "$newpid" -o comm=`
    sleep 0.1
  done**
  sleep 2
  # Check if the process has died; in that case we'll tail the log so the user can see
  if [[ ! $(ps -p "$newpid" -o comm=) =~ "java" ]]; then
    echo "failed to launch $command:"
    tail -2 "$log" | sed 's/^/  /'
    echo "full log in $log"
  fi
`

then run the start daemon script it will print such as:

sbin/start-master.sh 
starting org.apache.spark.deploy.master.Master, logging to /diskext/mySpark/spark/logs/spark-root-org.apache.spark.deploy.master.Master-1-n131.out
check daemon stage 1 bash
check daemon stage 2 bash
check daemon stage 3 bash
check daemon stage 4 bash
check daemon stage 5 java
check daemon stage 6 java
check daemon stage 7 java
check daemon stage 8 java
....

The running result show that the bash status daemon will exist for sometime.
Especially when the machine just startup and OS cache is clean,  the time may exceed 2s.

**Comment by srowen** [comment]
OK maybe, though here you show it took 0.5s.

File: sbin/spark-daemon.sh (hunk 2 above)
Comment (WeichenXu123) [inline sbin/spark-daemon.sh]: Oh..the sleep 2 statement can't move.

the loop may take at most 5 seconds to check whether the daemon turn into java daemon. When it become java daemon, it jump out the loop.

**then must sleep 2 seconds,**
then check whether the java daemon die. Because java daemon may exist for some time and then throw an Exception and die. So here need to wait for some time.

Comment (srowen) [inline sbin/spark-daemon.sh]: OK, that's fair, we still need to see if it dies. I suppose there could be a second loop to keep watching for it to die, but in the normal case that still requires waiting some number of seconds. This is OK, except for the PR problem of course.

Comment (WeichenXu123) [inline sbin/spark-daemon.sh]: new clean PR is here:
https://github.com/apache/spark/pull/13172
thanks.
